README.md
# Autonics LSC ROS Driver
This ROS Driver is for Autonics LSC-C Series

### Table of Contents

- [Supported Hardware](#1-supported-hardware)
- [ROS API](#2-ros-api)
- [Build](#3-build)
- [Start](#4-start)


### 1. Supported Hardware
#### 1.1 Model name : LSC-C Series
#####   Website : [Autonics/LiDAR/LSC-C](https://www.autonics.com/series/3001018)


### 2. ROS API
#### 2.1 Published Topics
* scan(sensor_msgs/LaserScan) : Scan data from the device
* diagnostics(diagnostic_updater) : Scan topic status
#### 2.2. Services
* self_test(self_test::Testrunner) : checking communication connection
#### 2.3. Parameters
* addr(default : 192.168.0.1, type : string) - The device ip address
* port(default : 8000, type : int) - The port number of device
* frame_id(default : laser, type : string) - The frame name of scan data
* range_min(default : 0.05, type : double) - Minimum range value [m]
* range_max(default : 25.0, type : double) - Maximum range value [m]
* password(default : 0000, type : string) - Password to login LSC-C


### 3. Build
####   3.1 from source
    source /opt/ros/<rosdistro>/setup.bash
    makdir -p ~/catkin_ws/src/
    cd ~/catkin_ws/src/
    git clone https://github.com/AutnoicsLiDAR/lsc_ros_driver.git
    cd ~/catkin_ws
    catkin_make


### 4. Start
    roslaunch lsc_ros_driver lsc_c25_launch.launch


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package lsc_ros_driver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1.0.2 (2022-07-26)
-----------
* Change pacakge info


1.0.1 (2022-07-19)
-----------
* Initial release


src/parser.cpp
/*
* Copyright (c) 2022, Autonics Co.,Ltd.
* All rights reserved.
* 
* Redistribution and use in source and binary forms, with or without 
* modification, are permitted provided that the following conditions
* are met:
* 
*   * Redistributions of source code must retain the above copyright 
*     notice, this list of conditions and the following disclaimer.
*
*   * Redistributions in binary form must reproduce the above 
*     copyright notice, this list of conditions and the following 
*     disclaimer in the documentation and/or other materials provided 
*     with the distribution.
*
*   * Neither the name of the Autonics Co.,Ltd nor the names of its 
*     contributors may be used to endorse or promote products derived 
*     from this software without specific prior written permission.
* 
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
* POSSIBILITY OF SUCH DAMAGE.
*/


#include <queue>
#include "parser.hpp"

#define NUM_OF_CMD_TYPE 4
#define NUM_OF_CMD 7

#define NUMBER_OF_PARAM 13
#define MAX_NUM_OF_PACKET_PARAM 16

#define INTENSITY_MAX 50000.0


const char* cmd_type_list[NUM_OF_CMD_TYPE] = 
{"sMC",
"sMA",
"sRC",
"sRA",
};

const char* cmd_list[NUM_OF_CMD] = 
{"None",
"SetAccessLevel",
"SensorScanInfo",
"SensorStart",
"SensorStop",
"ScanData",
"FirstConnectDummySend"
};

enum CmdListNum 
{
    NONE = 0,
    SET_ACCESS_LEVEL,
    SENSOR_SCAN_INFO,
    SENSOR_START,
    SENSOR_STOP,
    SCAN_DATA,
    FIRST_CONNECT_DUMMY_SEND
};

Parser::Parser()
{

}

int Parser::makeCommand(unsigned char* buf, std::string cmd)
{
    std::vector<unsigned char> v_cmd_frame;
    int cnt = 0, cmd_num = 0;
    std::string target = " ";
    std::string rep = ",";
    std::string temp_str;

    // replacing " " to ","
    for(int i = 0; cmd.find(target) != std::string::npos && i < MAX_NUM_OF_PACKET_PARAM; i++)
    {
        cmd.replace(cmd.find(target), target.length(), rep);
    }

    temp_str = cmd;
    char* ptr_tok_cmd = strtok(&temp_str[0], ",");

    // searching matched cmd
    for(int i = 0; i < NUM_OF_CMD; i++)
    {
        if(strcmp(cmd_list[i], ptr_tok_cmd) == 0)
        {
            cmd_num = i;
            break;  
        }
    }

    v_cmd_frame.push_back(',');

    switch(cmd_num)
    {
        case SET_ACCESS_LEVEL :
        case SENSOR_START :
        case SENSOR_STOP :
            v_cmd_frame.push_back('s');
            v_cmd_frame.push_back('M');
            v_cmd_frame.push_back('C');
            break;
        
        case SENSOR_SCAN_INFO :
            v_cmd_frame.push_back('s');
            v_cmd_frame.push_back('R');
            v_cmd_frame.push_back('C');
            break;

        case NONE :
        default :
            std::cout << "invalid command, type correct command " << cmd_num << std::endl;
            return -1; 
            break;
    }
    v_cmd_frame.push_back(',');

    for(int i = 0; i < cmd.size() + 1; i++)
    {
        v_cmd_frame.push_back(cmd[i]);
    }

    memcpy(buf, &v_cmd_frame[0], v_cmd_frame.size());

    return v_cmd_frame.size() - 1;
}

void Parser::parsingMsg(std::vector<unsigned char> raw_msg, sensor_msgs::LaserScan::Ptr msg, Lsc_t* lsc)
{
    uint16_t msg_size = 0, index= 0, field_size = 0, index_base = 0;
    int cmd_type = 0, cmd = 0, version = 0, status = 0;
    float chk_intensity = 0;
    std::vector<char *> field;
    char* ptr;
    
    if(raw_msg[0] == 0x02)
    {
        ptr = strtok((char*) &raw_msg[1], ",");

        if(raw_msg[strtoul(ptr,NULL, 16) - 1] == 0x03)
        {
            raw_msg[strtoul(ptr, NULL, 16) - 1] = '\0';
        }
        else
        {
            for(int i = 0; i < raw_msg.size(); i++)
            {
                std::cout << raw_msg[i] <<std::endl;
            }
            return;
        }

        while(ptr != NULL)
        {
            field.push_back(ptr);
            ptr = strtok(NULL, ",");
        }
    
        // get packet size
        msg_size = strtoul(field[index++], NULL, 16);
        field_size = field.size();

        // get command type
        for(int i = 0; i < NUM_OF_CMD_TYPE; i++)
        {
            int temp_ret = strcmp(field[index], cmd_type_list[i]);

            if(temp_ret == 0)
            {
                cmd_type = i;
                break;
            }
        }
        index++;

        // get command
        for(int i = 0; i < NUM_OF_CMD; i++)
        {
            int temp_ret = strcmp(field[index], cmd_list[i]);

            if(temp_ret == 0)
            {
                cmd = i;
                break;
            }
        }
        index++;

        switch(cmd)
        {
            case SENSOR_SCAN_INFO :
                lsc->scan_info.fw_ver = strtoul(field[8], NULL, 16);         // fw_ver
                lsc->scan_info.model_name = field[13];                        // model name
                
                #ifdef PRINT_PARSE_DATA                
                std::cout << "Command : " << cmd_list[cmd] << std::endl;
                std::cout << "fw_ver : " << lsc->scan_info.fw_ver << std::endl;
                std::cout << "Model_name : " << lsc->scan_info.model_name << std::endl;
                #endif

                break;

            case SCAN_DATA :
                for(int i = 0; i < NUMBER_OF_PARAM; i++)
                {
                    ptr = strtok(NULL, ",");
                    field.push_back(ptr);
                }
                lsc->scan_mea.scan_counter = strtoul(field[7], NULL, 16);       // scan counter
                lsc->scan_mea.scan_freq = strtoul(field[10], NULL, 16);          // scan freq
                lsc->scan_mea.meas_freq = strtoul(field[11], NULL, 16);          // meas freq
                lsc->scan_mea.angle_begin = strtoul(field[12], NULL, 16);        // angle_begin
                lsc->scan_mea.angle_resol = strtoul(field[13], NULL, 16);        // angle_resol
                lsc->scan_mea.amnt_of_data = strtoul(field[14], NULL, 16);       // amount of data
                index = 16;

                #ifdef PRINT_PARSE_DATA                
                std::cout << "Cmd type : " << cmd_type_list[cmd_type] << std::endl;
                std::cout << "Command : " << cmd_list[cmd] << std::endl;
                std::cout << "Scan counter : " << lsc->scan_mea.scan_counter << std::endl;
                std::cout << "Scan frequency : " << std::dec << lsc->scan_mea.scan_freq << std::endl;
                std::cout << "Measuring frequency : " << lsc->scan_mea.meas_freq << std::endl;
                std::cout << "Angle begin : " << lsc->scan_mea.angle_begin << std::endl;
                std::cout << "Angle resolution : " << lsc->scan_mea.angle_resol << std::endl;
                std::cout << "Amount of data : " << lsc->scan_mea.amnt_of_data << std::endl;
                #endif

                msg->angle_min = lsc->scan_mea.angle_begin / 10000.0 * DEG2RAD;
                msg->angle_max = (lsc->scan_mea.angle_begin / 10000.0 + (lsc->scan_mea.angle_resol / 10000.0 * lsc->scan_mea.amnt_of_data)) * DEG2RAD;
                msg->angle_increment = lsc->scan_mea.angle_resol / 10000.0 * DEG2RAD;
                msg->time_increment = (60.0 / lsc->scan_mea.meas_freq) / lsc->scan_mea.scan_counter;
                msg->scan_time = (60.0 / lsc->scan_mea.meas_freq); 

                #ifdef PRINT_PARSE_DATA 
                std::cout << "" << std::endl;
                std::cout << "msg->angle_min = " << msg->angle_min << std::endl;
                std::cout << "msg->angle_max = " << msg->angle_max << std::endl;
                std::cout << "msg->angle_increment = " << msg->angle_increment << std::endl;
                std::cout << "msg->range_max = " << msg->range_max << std::endl;
                std::cout << "msg->range_min = " << msg->range_min << std::endl;
                std::cout << "msg->scan_time = " << msg->scan_time << std::endl;
                std::cout << "msg->time_increment = " << msg->time_increment << std::endl;
                #endif
                
                if(strcmp(field[index++], "DIST1") == 0)
                {
                    index_base = index;

                    msg->ranges.clear();
                    for(uint16_t i = 0; i < lsc->scan_mea.amnt_of_data; i++)
                    {
                        msg->ranges.push_back(strtoul(field[index++], NULL, 16) / 1000.0);
                        
                    #ifdef PRINT_PARSE_DATA
                        std::cout << msg->ranges[i] << ", ";
                    }
                    std::cout << "" << std::endl;
                    #else
                    }                   
                    #endif
                }

                msg->intensities.clear();
                if(field[index] == field.back())
                {   
                }
                else
                {
                    if(strcmp(field[index++], "RSSI1") == 0 )
                    {
                        index_base = index;

                        for(uint16_t i = 0; i < lsc->scan_mea.amnt_of_data; i++)
                        { 
                            chk_intensity = strtoul(field[index++], NULL, 16);                            

                            msg->intensities.push_back(chk_intensity);

                        #ifdef PRINT_PARSE_DATA
                            std::cout << std::hex << std::uppercase << msg->intensities[i] << ",";
                        }
                        std::cout << "" << std::endl;
                        #else
                        }
                        #endif
                    }
                } 
                break;

            case SENSOR_START :
            case SENSOR_STOP :
            case SET_ACCESS_LEVEL :
            case FIRST_CONNECT_DUMMY_SEND :
                for(int i = 2; i < field.size(); i++)
                {
                    std::cout << field[i] << " ";
                }
                std::cout << "" << std::endl;
                break;
                
            default :
                for(int i = 2; i < field.size(); i++)
                {
                    std::cout << field[i] << " ";
                }
                std::cout << "" << std::endl;
                break;
        }
    }
}

src/socket.cpp
/*
* Copyright (c) 2022, Autonics Co.,Ltd.
* All rights reserved.
* 
* Redistribution and use in source and binary forms, with or without 
* modification, are permitted provided that the following conditions
* are met:
* 
*   * Redistributions of source code must retain the above copyright 
*     notice, this list of conditions and the following disclaimer.
*
*   * Redistributions in binary form must reproduce the above 
*     copyright notice, this list of conditions and the following 
*     disclaimer in the documentation and/or other materials provided 
*     with the distribution.
*
*   * Neither the name of the Autonics Co.,Ltd nor the names of its 
*     contributors may be used to endorse or promote products derived 
*     from this software without specific prior written permission.
* 
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
* POSSIBILITY OF SUCH DAMAGE.
*/


#include <sys/poll.h>
#include <stdio.h>

#include "socket.hpp"

Socket::Socket()
{
    m_server_sock_ = 0;
    m_connected_ = 0;

    rcv_timeout_ = false;
    rcv_error_ = false;

    pthread_running_ = false;
    pthrd_id_ = 0;
}

Socket::~Socket()
{
    pthread_running_ = false;

    if(pthrd_id_)
    {
        pthread_join(pthrd_id_, NULL);
    }

    if(m_connected_ == true)
    {
        close(m_server_sock_);

        m_connected_ = false;

        std::cout << "disconnect server" << std::endl;
    }
}

int Socket::getServerSocket(void)
{
    return m_server_sock_;
}

bool Socket::getConnected(void)
{
    return m_connected_;
}

bool Socket::getPthreadRunning(void)
{
    return pthread_running_;
}


bool Socket::getRcvTimeout(void)
{
    return rcv_timeout_;
}

bool Socket::getRcvError(void)
{
    return rcv_error_;
}

void Socket::setRcvTimeout(bool flag)
{
    rcv_timeout_ = flag;
}

void Socket::setRcvError(bool flag)
{
    rcv_error_ = flag;
}



int Socket::tryReconnection(void)
{
    int ret = 0;
    pthread_running_ = false;

    if(pthrd_id_)
    {
        pthread_join(pthrd_id_, NULL);
        pthrd_id_ = 0;
    }
    
    m_connected_ = false;
    close(m_server_sock_);

    usleep(1000*1000);

    ret = clientOpen(addr_st_, port_num_st_);

    if(ret < 0)
    {
        std::cout << "reconnection failed" << std::endl;
        return ret;
    }
    
    return ret;
}

void Socket::putBufToMsg(unsigned char* buf, uint16_t size)
{
    std::vector<unsigned char> vec;

    for(int i = 0; i < size; i++)
    {
        vec.push_back(*(buf+i));
    }

    recvQueue.push(vec);   
}


void* readCallback(void* arg)
{
    int32_t read_cnt = 0;
    uint16_t packet_size = 0, total_byte = 0, offset = 0, recv_byte = 0, \
             buf_size = 16384;
    int ret = 0;
    unsigned char r_buffer[buf_size] = {0, }, \
                  buffer[buf_size] = {0, }, \
                  cp_temp[16] = {0, };
    Socket* sock = (Socket*) arg;
    struct pollfd fd;
    ssize_t chk_connection = 0;

    fd.fd = sock->getServerSocket();
    fd.events = POLLIN;

    while(sock->getPthreadRunning())
    {
        read_cnt = 0;
        offset = 0;

        ret = poll(&fd, 1, 1000);

        if(ret < 0)
        {
            sock->setRcvError(true);
            std::cout << "polling error" << std::endl;
        }
        else if(ret == 0)
        {
            read_cnt = recv(sock->m_server_sock_, r_buffer, sizeof(r_buffer), 0);
            if(read_cnt < 0)
            {
                *r_buffer = 'a';
                chk_connection = write(sock->m_server_sock_, r_buffer, sizeof(r_buffer));

                if(chk_connection == -1)
                {
                    sock->setRcvTimeout(true);
                    std::cout << "server disconnected" << std::endl;
                }
            }
        }
        else // ret > 0
        {
            sock->setRcvError(false);

            read_cnt = sock->clientRead(r_buffer, sizeof(r_buffer));            
            if(read_cnt > 0)
            {      
                sock->setRcvTimeout(false);
                // Search packet size
                if(packet_size == 0)
                {
                    for(uint16_t i = 0; i < read_cnt; i++)
                    {
                        if(r_buffer[i] == 0x02)
                        {
                            for(int j = 0; j < 4; j++)
                            {
                                cp_temp[j] = r_buffer[i+1+j];
                            }
                            packet_size = strtoul((const char*) cp_temp, NULL, 16);
                            break;
                        }
                    }
                }

                // Search ETX
                for(uint16_t i = 0; i < read_cnt - 1; i++)
                {
                    if(r_buffer[read_cnt - 1 - i] == 0x03)
                    {
                        read_cnt -= i;
                        break;
                    }
                }
                
                if(packet_size > total_byte)
                {
                    memcpy(buffer + total_byte, r_buffer + offset, read_cnt - offset);
                    total_byte += read_cnt - offset;
                }   
                else if(packet_size < total_byte)
                {
                    if(buffer[0] == 0x02 && buffer[packet_size] == 0x03)
                    {
                        total_byte = packet_size; 
                    }
                    else
                    {
                        memset(buffer, 0x00, sizeof(buffer));
                        packet_size = 0;
                        total_byte = 0;
                        offset = 0;
                    }
                }
                else
                {

                }

                if(packet_size !=0 && total_byte != 0 && packet_size == total_byte)
                {
                    if(buffer[total_byte - 1] == 0x03)
                    {
                        // push in the queue
                        sock->putBufToMsg(buffer, total_byte);
                        
                        memset(buffer, 0x00, sizeof(buffer));
                        packet_size = 0;
                        total_byte = 0;
                    }
                }
                else
                {

                }                
                memset(r_buffer, 0x00, sizeof(r_buffer));
            }
            else
            {
                sock->setRcvTimeout(true);
                std::cout << "server response timeout" << std::endl;
            }
        }
    }

    return 0;
}


int Socket::clientOpen(std::string addr, std::string port)
{
    int m_server_addr_size = 0;
    uint16_t port_num = 0;

    timeval tv;
    
    
    addr_st_ = addr;
    port_num_st_ = port;
    
    m_server_sock_ = socket(PF_INET, SOCK_STREAM, 0);    

    if(m_server_sock_ < 0)
    {
        std::cout << "faild to create client socket" << std::endl;
        return -1;
    }

    tv.tv_sec = 0;;
    tv.tv_usec = 500*1000;

    if(setsockopt(m_server_sock_, SOL_SOCKET, SO_RCVTIMEO, (char*)&tv, sizeof(tv)) < 0)
    {
        std::cout << "failed rcvtimeo setsockopt" << std::endl;
    }
    if(setsockopt(m_server_sock_, SOL_SOCKET, SO_SNDTIMEO, (char*)&tv, sizeof(tv)) < 0)
    {
        std::cout << "failed sndtimeo setsockopt" << std::endl;
    }

    sscanf(port.c_str(), "%u", (unsigned int*)&port_num);

    memset(&m_server_addr_, 0x00, sizeof(m_server_addr_));

    m_server_addr_.sin_family = AF_INET;
    m_server_addr_.sin_addr.s_addr = inet_addr(addr.begin().base());
    m_server_addr_.sin_port = htons(port_num);

    m_server_addr_size = sizeof(m_server_addr_);
    
    if(connect(m_server_sock_, (const sockaddr*)&m_server_addr_, m_server_addr_size) < 0)
    {
        std::cout << "failed client connect, " << strerror(errno) << std::endl;
        return -1;
    }
    else
    {
        m_connected_ = true;
    }

    if(pthread_create(&pthrd_id_, NULL, readCallback, this) < 0)
    {
        std::cout << "pthread create failed" << std::endl;
        return -1;
    }
    else
    {
        pthread_running_ = true;
    }    

    return 0;
}

int32_t Socket::clientRead(unsigned char* buffer, uint16_t buf_size)
{
    int32_t read_size = 0;

    if(m_connected_ == true)
    {
       read_size = recv(m_server_sock_, buffer, buf_size, 0);

       if(read_size < 0)
        {
            std::cout << "client read failed" << std::endl;
        }
        
        return read_size;
    }
    else
    {
        std::cout << "server is disconnected" << std::endl;
        return -1;
    }
}

int32_t Socket::clientWrite(unsigned char* buffer, uint16_t buf_size)
{
    int32_t write_size = 0;

    if(m_connected_ == true)
    {
        write_size = send(m_server_sock_, buffer, buf_size, 0);
        if(write_size < 0)
        {
            std::cout << "client write failed" << std::endl;

            try
            {

            }
            catch(const std::exception& e)
            {
                std::cerr << e.what() << '\n';
            }
            
        }
        return write_size;
    }
    else
    {
        std::cout << "server is disconnected" << std::endl;
        return -1;
    }

    
}



src/laser.cpp
/*
* Copyright (c) 2022, Autonics Co.,Ltd.
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions
* are met:
*
*   * Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*
*   * Redistributions in binary form must reproduce the above
*     copyright notice, this list of conditions and the following
*     disclaimer in the documentation and/or other materials provided
*     with the distribution.
*
*   * Neither the name of the Autonics Co.,Ltd nor the names of its
*     contributors may be used to endorse or promote products derived
*     from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/




#include <iostream>
#include <sstream>

#include <self_test/self_test.h>

#include "laser.h"

AutoLaser::AutoLaser() : n(), priv_nh("~"), scan_msg(new sensor_msgs::LaserScan)
{
    rcv_msg_flag_ = false;
}

AutoLaser::~AutoLaser()
{

}

std::string AutoLaser::itostr(int i)
{
    std::string rt_str;
    std::stringstream ss;

    ss << i;
    ss >> rt_str;

    return rt_str;
}

std::string AutoLaser::ftostr(float i)
{
    std::string rt_str;
    std::stringstream ss;

    ss << i;
    ss >> rt_str;

    return rt_str;
}

unsigned char AutoLaser::hexToChar(unsigned int value)
{
    if(value > 9 && value < 16)
    {
        return value + '0' + 7;
    }
    else if (value >= 16)
    {
        ROS_WARN("the value is out of range");
        return value;
    }
    else
    {
        return value + '0';
    }
}

void AutoLaser::comSubCallback(const std_msgs::String::ConstPtr& msg)
{
    std::string cmd = msg->data.c_str();
    laserSendCommand(cmd);
}


void AutoLaser::login(std::string password)
{
    std::string login_cmd = "SetAccessLevel";
    login_cmd = login_cmd + " " + password;

    if(laserSendCommand(login_cmd) < 0);
}

int AutoLaser::laserSendCommand(const std::string str)
{
    uint16_t s_buffer_size = 0;
    unsigned char s_buffer[64] = {0, };
    int cmd_str_cnt = 0, str_cnt = 0;

    memset(s_buffer, 0x00, sizeof(s_buffer));

    s_buffer[str_cnt++] = 0x02;
    str_cnt++;
    str_cnt++;
    str_cnt++;
    str_cnt++;

    cmd_str_cnt = p.makeCommand(&s_buffer[str_cnt], str);

    if(cmd_str_cnt == -1)
    {
        return -1;
    }
    else
    {
        str_cnt += cmd_str_cnt;
    }
    s_buffer[str_cnt++] = 0x03;

    s_buffer[1] = hexToChar(str_cnt / 16 / 16 / 16);
    s_buffer[2] = hexToChar(str_cnt / 16 / 16 % 16);
    s_buffer[3] = hexToChar(str_cnt / 16 % 16);
    s_buffer[4] = hexToChar(str_cnt % 16);

    s_buffer_size = sizeof(s_buffer);

    if(socket.clientWrite(s_buffer, s_buffer_size) < 0)
    {
        ROS_ERROR("Client write failed");
        if(socket.tryReconnection() < 0)
        {
            return -1;
        }
        else
        {
            ROS_INFO("Reconnect successfully");
        }
    }
    else
    {
        #ifdef PRINT_PARSE_DATA
        std::cout << std::endl << "sent data : ";
        for(int i = 0; i < str_cnt; i++)
        {
            std::cout << std::hex << s_buffer[i];
        }
        std::cout << "" << std::endl;
        #endif

        usleep(1000*100);
    }

    return 0;
}

bool AutoLaser::checkConnection()
{
    bool result =  false;

    if(socket.getRcvError() || socket.getRcvTimeout())
    {

    }
    else
    {
        result = true;
    }

    return result;
}

void AutoLaser::selfTest(diagnostic_updater::DiagnosticStatusWrapper& status)
{
    status.hardware_id = lsc.scan_info.model_name;

    if(!checkConnection())
    {
        status.summary(diagnostic_msgs::DiagnosticStatus::ERROR, "lost connection");
    }
    else
    {
        status.summary(diagnostic_msgs::DiagnosticStatus::OK, "connection is ok");
    }
}

int AutoLaser::getLscData(void)
{
    if(!(socket.recvQueue.empty()))
    {
        rcv_msg_ = socket.recvQueue.front();
        socket.recvQueue.pop();

        p.parsingMsg(rcv_msg_, scan_msg, &lsc);
    }
    else
    {
        return -1;
    }

    return 0;
}

int AutoLaser::getLaserData(sensor_msgs::LaserScan::Ptr p_laser_scan)
{
    if(!(socket.recvQueue.empty()))
    {
        rcv_msg_ = socket.recvQueue.front();
        socket.recvQueue.pop();

        p.parsingMsg(rcv_msg_, p_laser_scan, &lsc);
        if(p_laser_scan->ranges.empty() == false)
        {
            p_laser_scan->header.stamp = ros::Time::now();

            rcv_msg_flag_ = true;
        }
    }
    else
    {
        rcv_msg_flag_ = false;
    }

    return 0;
}

void AutoLaser::watchingDisconnection(void)
{
    if(socket.getRcvTimeout())
    {
        if(socket.tryReconnection() >= 0)
        {
            ROS_INFO("Reconnect successfully");
            laserSendCommand("SensorStart");
        }
        else
        {

        }
    }
}

int AutoLaser::laserInit(void)
{
    std::string addr = "192.168.0.1", frame_id = "laser", port = "8000", password = "0000";
    double range_min = 0.05, range_max = 25;
    uint16_t port_num = 0, recnt_cnt = 0;

    pub_scan = n.advertise<sensor_msgs::LaserScan>("scan", 1000);
    sub_cmd = n.subscribe<std_msgs::String>("command", 10, &AutoLaser::comSubCallback, this);

    priv_nh.getParam("addr", addr);
    priv_nh.getParam("frame_id", frame_id);
    priv_nh.getParam("port", port);
    priv_nh.getParam("range_min", range_min);
    priv_nh.getParam("range_max", range_max);
    priv_nh.getParam("password", password);

    scan_msg->header.frame_id = frame_id;
    scan_msg->range_min = range_min;
    scan_msg->range_max = range_max;

    if(socket.clientOpen(addr, port) < 0)
    {
        ROS_WARN("Failed to connect to server");
        while(true)
        {
            recnt_cnt++;
            if(socket.tryReconnection() < 0)
            {
                ROS_WARN("Failed to reconnect to server");
                if(recnt_cnt >= CONNECT_ATTEMPT_CNT)
                {
                    ROS_ERROR("Tried to connect server %d times but, Failed to connect to server", recnt_cnt);
                    return -1;
                }
                if(!ros::ok())
                {
                    return -1;
                }
            }
            else
            {
                break;
            }
        }
    }

    laserSendCommand("SensorScanInfo");
    getLscData();

    login(password);

    return 0;
}

int AutoLaser::run(void)
{
    self_test::TestRunner self_test;

    self_test.setID("self_test");
    self_test.add("Connection", this, &AutoLaser::selfTest);

    double min_freq = 0.00025;  // Hz
    double max_freq = 15.0;     // Hz
    diagnostic_updater::HeaderlessTopicDiagnostic pub_freq("scan", diagnostic_topic_updater, diagnostic_updater::FrequencyStatusParam(&min_freq, &max_freq, 0.1, 1));
    diagnostic_topic_updater.setHardwareID("AutoLaser");

    laserSendCommand("SensorStart");

    while(ros::ok())
    {
        getLaserData(scan_msg);

        if(rcv_msg_flag_)
        {
            pub_scan.publish(scan_msg);
            scan_msg->ranges.clear();
            scan_msg->intensities.clear();

            pub_freq.tick();
            diagnostic_topic_updater.update();
        }

        self_test.checkTest();

        watchingDisconnection();

        ros::spinOnce();
    }

    return 0;
}

